### PR TITLE
Stuff

### PR DESCRIPTION
**Pull request by brentc**
Fixes #1

### CODE REVIEW

**Comment by marz619** [comment]
Please provide a better title and a comprehensive description of what is in the commit

File: cart.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import sqlite3
+
+class HTTPResponse(object):
```
Comment (marz619) [inline cart.py]: Move this class to a separate file since it is not directly related to the Cart

File: cart.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        ``.discounts``: a list of item-pair discounts that apply to this cart
+            in the form ``(discount_name, discount_amount)``.
+
+        ... additional properties omitted ...
+    """
+
+    # ... omitted ...
+
+
+class Template(object):
```
Comment (marz619) [inline cart.py]: Move this class to a separate file since it is not directly related to the Cart. Maybe `template.py`

File: cart.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        in special deals.
+
+        Returns an "added to cart" page, confirming the name of the item which
+        has been added, as well as the current quantity and messages about
+        any special discounts. """
+
+    def __init__(self, settings):
+        self.settings = settings
+
+    def handle_GET(self, request):
```
Comment (marz619) [inline cart.py]: Since this is making a change to the data on the server side, it should be handled as a `POST` request

File: cart.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            a soccer ball and goalie gloves, the "goal keeper" discount
+            would be applied).
+
+            Returns a description of the discounts, or `""` if no
+            discounts apply. """
+        cxn = sqlite3.connect(self.settings.DATABASE_PATH)
+        discounts = []
+        for item0 in cart.items:
+            for item1 in cart.items:
+                result = cxn.execute("""
```
Comment (marz619) [inline cart.py]: Should be using a `prepared statement` and bind any arguments instead of directly formatting the select statement.

File: cart.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            return """
+                This shopping cart is eligible for the following discounts: %s<br>
+                For a total savings of $%s.
+            """ %(discounts_str, total_discount)
+        return ""
+
+    def __validate_item_id(self, item_id):
+        try:
+            cxn = sqlite3.connect(self.settings.DATABASE_PATH)
+            result = cxn.execute("SELECT * FROM items WHERE id = " + item_id)
```
Comment (marz619) [inline cart.py]: Should be using a `prepared statement` here.

File: cart.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+        # Update the quantity of that item in the cart
+        current_quantity = cart.item_quantities.get(item_id, 0)
+        current_quantity += int(request.GET.get("quantity"))
+
+        # Check to see if the cart is eligible for any special discounts
+        special_discount_message = self.__check_cart_for_discounts(cart)
+
+        # Send back the response
+        response_body = Template("""
```
Comment (marz619) [inline cart.py]: Move the template rendering into it's own method to reduce the overall size of the `handler` method

File: cart.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                itemIndex = i
+                break
+
+        if itemIndex < 0:
+            itemIndex = len(cart.items)
+            cart.items.append(item_id)
+
+        # Update the quantity of that item in the cart
+        current_quantity = cart.item_quantities.get(item_id, 0)
+        current_quantity += int(request.GET.get("quantity"))
```
Comment (marz619) [inline cart.py]: `int()` will throw an Exception if the `quantity` parameter is not a valid number. Need to handle the error here.

File: cart.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        try:
+            cxn = sqlite3.connect(self.settings.DATABASE_PATH)
+            result = cxn.execute("SELECT * FROM items WHERE id = " + item_id)
+            result = result.fetchone()
+            cxn.close()
+            if result == None:
+                return False
+            else:
+                return True
+        except:
```
Comment (marz619) [inline cart.py]: Could do more explicit error handling here. In order to return a more comprehensive response to the client, and for error logging purposes.

File: cart.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                """ %(item0, item1))
+                result = result.fetchone()
+                if result == None:
+                    continue
+                discounts.append(result)
+        cart.discounts = discounts
+
+        discounts_str = ""
+        total_discount = 0
+        for name_amount in cart.discounts:
```
Comment (marz619) [inline cart.py]: potentially NULL values returned from the database could cause an error here.

File: cart.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        response_body = Template("""
+            <div class='addToCartConfirmation'>
+                The item {{ item_name }} has been added to your cart.<br />
+                There are {{ quantity }} of them in your cart.
+            </div>
+            <div class='specialDiscountMessage'>
+                {{ special_discount_message }}
+            </div>
+        """).render({
+            "item_name": request.GET["item_name"],
```
Comment (marz619) [inline cart.py]: what if "item_name" is not in the request object.

File: cart.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(self, settings):
+        self.settings = settings
+
+    def handle_GET(self, request):
+        item_id = request.GET.get("item_id")
+        if not self.__validate_item_id(item_id):
+            return HTTPResponse(500, "No such item with id '" + item_id + "' does not exist.")
+
+        cart = request.session.cart
```
Comment (marz619) [inline cart.py]: Is the `cart` object guaranteed to exist in the session object?

File: cart.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        has been added, as well as the current quantity and messages about
+        any special discounts. """
+
+    def __init__(self, settings):
+        self.settings = settings
+
+    def handle_GET(self, request):
+        item_id = request.GET.get("item_id")
+        if not self.__validate_item_id(item_id):
+            return HTTPResponse(500, "No such item with id '" + item_id + "' does not exist.")
```
Comment (marz619) [inline cart.py]: This response in not necessarily true if there was a connection error in retrieving the data from database.

**Comment by marz619** [comment]
It might be prudent to separate the SQL connection handling into an explicit Connection Pool if available, and also create a Data Access Object for the Cart with methods to perform various actions / having the prepared statements etc. contained within, as adding an item to a cart is not the only thing we will be doing.

**Comment by robynsmith** [comment]
A couple of notes:

- There were not tests written for this new code.
- def __validate_item_id(self, item_id):
     - This function does not attempt to validate the ID before performing a SQL query, which could cause portability and/or other issues with databases.
- This section:
Line 116: discounts.append(result)

Could potentially result in unintended side-effects, as it is appending “None” to the end of discounts.

For the most part, though, looks like a solid concept for a shopping cart!

File: cart.py (hunk 12 above)
Comment (ldom) [inline cart.py]: I would either use a 400 or a 404 code here.

File: cart.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        self.mimetype = mimetype
+        # ... additional properties omitted ...
+
+    # ... omitted ...
+
+
+class Cart(object):
+    """ A user's shopping cart.
+        
+        ``.items``: a list of item IDs in the cart.
```
Comment (ldom) [inline cart.py]: Using a list for the cart items doesn't provide for a fast and legible way to find the items later in the code. I'd use a dict with the item id as the key.

File: cart.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    def handle_GET(self, request):
+        item_id = request.GET.get("item_id")
+        if not self.__validate_item_id(item_id):
+            return HTTPResponse(500, "No such item with id '" + item_id + "' does not exist.")
+
+        cart = request.session.cart
+
+        # Find the item in the cart
+        itemIndex = -1
+        for i in range(len(cart.items)):
```
Comment (ldom) [inline cart.py]: See remark on line 16. To be replaced by cart.items[item_id] enclosed in a try/except KeyError.

File: cart.py (hunk 4 above)
Comment (ldom) [inline cart.py]: The 2 for loops and the query should be replaced by a more efficient construct. One way would be to build a dict with item_ids being part of discounts as keys, and intersect the results.

File: cart.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        cxn = sqlite3.connect(self.settings.DATABASE_PATH)
+        discounts = []
+        for item0 in cart.items:
+            for item1 in cart.items:
+                result = cxn.execute("""
+                    SELECT discount_name, discount_amount FROM item_discounts
+                    WHERE item0 = %s and item1 = %s
+                """ %(item0, item1))
+                result = result.fetchone()
+                if result == None:
```
Comment (ldom) [inline cart.py]: if result is None

**Comment by evert** [comment]
Not a great PR title and commit message ;)

File: cart.py (hunk 12 above)
Comment (evert) [inline cart.py]: 404 would be more appropriate here.

File: cart.py (hunk 14 above)
Comment (evert) [inline cart.py]: Could be better solved using `cart.items.index(item_id)`.

File: cart.py (hunk 7 above)
Comment (evert) [inline cart.py]: `current_quantity` is not assigned to anything. Did you mean to set it back to cart.item_quantities?

File: cart.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+            would be applied).
+
+            Returns a description of the discounts, or `""` if no
+            discounts apply. """
+        cxn = sqlite3.connect(self.settings.DATABASE_PATH)
+        discounts = []
+        for item0 in cart.items:
+            for item1 in cart.items:
+                result = cxn.execute("""
+                    SELECT discount_name, discount_amount FROM item_discounts
```
Comment (evert) [inline cart.py]: If there's a 100 unique items in your cart, this would result in 100.000 sql queries. This should be optimized a bit into a single query. I also believe that this logic should really live in a separate class and not get intertwined with templates and controller-like functionality.

File: cart.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        if total_discount > 0:
+            return """
+                This shopping cart is eligible for the following discounts: %s<br>
+                For a total savings of $%s.
+            """ %(discounts_str, total_discount)
+        return ""
+
+    def __validate_item_id(self, item_id):
+        try:
+            cxn = sqlite3.connect(self.settings.DATABASE_PATH)
```
Comment (evert) [inline cart.py]: 1. This could also be a performance issue depending on how many items we want to support in a cart
2. This functionality should also move to a separate class
3. In that class, an instance of the sqlite3 Connection class should ideally be injected, and not constructed straight within the function.

**Comment by evert** [comment]
I would say that in general the biggest issues with this PR is:

1. You should really make smaller classes that have less responsibilities and learn about dependency injection.
2. Before you submit code for review, it has to be tested. There was at least one problem that lead me to believe that this code would not successfully run. To me this means at least a manual functional test by the developer, but preferably also automated tests for business/domain logic

**Comment by evert** [comment]
Note that I may have missed some details, but those are really major issues that need to be resolved before we can dive into them ;)
